### PR TITLE
Fixed Record Deployment

### PR DESCRIPTION
**Pull request by fciacchi**
- Developer(s): @fciacchi 

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   NewRelic  | major                 | RecordDeployment     |

#### Release Notes

Fixed bug in monitoring where console commands are not correctly sent to the monitoring service.

-----------------------------------------

##### Change log

Improvements

* Fixed correct console command logging in `spryker/monitoring` here: https://github.com/spryker/spryker/pull/5555 - potential release version `2.0.0`
* Added conflict constraint to previous core version of the new-relic module
* Fixed wrong GuzzleHttp namespace. Composer force version 6.x.x
* Fixed required arguments for the record deployment. Only `revision` (the 3rd) is required, therefore since the first argument was already required, made the first 3 arguments all required
* Fixed the way record deployment is set to NewRelic API: https://docs.newrelic.com/docs/apm/new-relic-apm/maintenance/record-deployments
* Improved the record deployment in a retro-compatible way. The record deployment URL can contain only one application-id (that is NewRelic term for an environment). Added application-id array, so that record deployment URL can have `sprintf` style placeholder and a list of IDs can be passed separately.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
   "license": "proprietary",
   "require": {
     "php": ">=7.1",
+    "spryker/monitoring": "^2.0.1",
```
Comment (stereomon) [inline composer.json]: Why is this required?

File: src/SprykerEco/Zed/NewRelic/Business/Model/RecordDeployment.php [hunk 2]
```diff
@@ -27,14 +31,25 @@ class RecordDeployment implements RecordDeploymentInterface
      */
     protected $newRelicApiKey;
 
+    /**
+     * @var array
+     */
+    protected $newRelicApplicationIdArray;
```
Comment (stereomon) [inline src/SprykerEco/Zed/NewRelic/Business/Model/RecordDeployment.php]: `$newRelicApplicationIds` or `$newRelicApplicationIdentifier`

File: src/SprykerEco/Zed/NewRelic/Business/Model/RecordDeployment.php [hunk 3]
```diff
@@ -46,17 +61,17 @@ public function __construct(string $newRelicDeploymentApiUrl, string $newRelicAp
...
-            throw new RecordDeploymentException(sprintf(
-                'Record deployment to New Relic request failed with code %d. %s',
-                $response->getStatusCode(),
-                $response->getBody()
-            ));
+        if (empty($this->newRelicApplicationIdArray)) {
+            return $this->recordSingleDeployment($arguments);
+        }
+
+        foreach ($this->newRelicApplicationIdArray as $singleApplicationId) {
```
Comment (stereomon) [inline src/SprykerEco/Zed/NewRelic/Business/Model/RecordDeployment.php]: Change variable names here as well.

File: src/SprykerEco/Zed/NewRelic/Business/Model/RecordDeployment.php [hunk 4]
```diff
@@ -66,23 +81,51 @@ public function recordDeployment(array $arguments = []): RecordDeploymentInterfa
...
+     * @throws \SprykerEco\Zed\NewRelic\Business\Exception\RecordDeploymentException
+     *
+     * @return $this
+     */
+    private function recordSingleDeployment(array $arguments=[]): RecordDeploymentInterface
+    {
+        $response = $this->createRecordDeploymentRequest($arguments);
+        $statusCode = $response->getStatusCode();
+        if ($statusCode < static::STATUS_CODE_SUCCESS || $statusCode >= static::STATUS_CODE_REDIRECTION) {
+            throw new RecordDeploymentException(sprintf(
```
Comment (stereomon) [inline src/SprykerEco/Zed/NewRelic/Business/Model/RecordDeployment.php:122]: Will that break the deployment? Shouldn't we only log here?

File: src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php [hunk 5]
```diff
@@ -53,13 +53,13 @@ protected function configure(): void
 
         $this->addArgument(
             self::ARGUMENT_USER,
-            InputArgument::OPTIONAL,
+            InputArgument::REQUIRED,
```
Comment (stereomon) [inline src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php:56]: Isn't it a breaking change?

File: src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php [hunk 6]
```diff
@@ -53,13 +53,13 @@ protected function configure(): void
...
             self::ARGUMENT_USER_DESCRIPTION
         );
 
         $this->addArgument(
             self::ARGUMENT_REVISION,
-            InputArgument::OPTIONAL,
+            InputArgument::REQUIRED,
```
Comment (stereomon) [inline src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php:62]: Same as above.

**Review by stereomon — CHANGES_REQUESTED** [review]

**Comment by stereomon** [comment]
Please add proper release template here as well.

File: composer.json (hunk 1 above)
Comment (fciacchi) [reply composer.json]: Monitoring is made of 3 parts: web, console commands and record deployment. Console command logging is broken in Core, therefore it's needed to force the right fixed version for the New Relic monitoring to work out of the box.

File: src/SprykerEco/Zed/NewRelic/Business/Model/RecordDeployment.php (hunk 4 above)
Comment (fciacchi) [reply src/SprykerEco/Zed/NewRelic/Business/Model/RecordDeployment.php:122]: This is the exactly same throw exception of here: https://github.com/spryker-eco/new-relic/pull/3/files#diff-58b5cba6df57adf679f01002327b7cb6L52
But, this might also run in a loop of IDs. I think it's the right behaviour anyway. Even with several applications, deployment shouldn't be an issue (it would be only if one of IDs is wrong, I guess).

File: src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php (hunk 6 above)
Comment (fciacchi) [reply src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php:62]: Revision is the only required parameter by new-relic API. Therefore if someone is using the record deployment, he still needs to send the first (required), the second (optional) and this (third one) to record a deployment. 
It's not a BC breaks, because it cannot work if we send only the first parameter anyway.

File: src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php (hunk 5 above)
Comment (fciacchi) [reply src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php:56]: As explained for the third argument. By having these arguments positional, I have to force the second one to be required as well

File: composer.json (hunk 1 above)
Comment (fciacchi) [reply composer.json]: I changed to 2.0.0 cause I cannot force a dependency that is not there. So, for both new and old users, by then updating this package, they will automatically fall on the spryker/monitoring 2.0.1 (unless they change and specify 2.0.0 manually, but that is unlikely)

File: src/SprykerEco/Service/NewRelic/Plugin/NewRelicMonitoringExtensionPlugin.php [hunk 7]
```diff
@@ -106,6 +106,7 @@ public function markIgnoreTransaction(): void
             return;
         }
 
+        \newrelic_ignore_apdex();
```
Comment (fciacchi) [inline src/SprykerEco/Service/NewRelic/Plugin/NewRelicMonitoringExtensionPlugin.php:109]: Additional command to ignore apdex for ignored transactions.

File: src/SprykerEco/Shared/NewRelic/NewRelicEnv.php [hunk 8]
```diff
@@ -24,4 +24,13 @@ interface NewRelicEnv
      * @api
      */
     const NEW_RELIC_API_KEY = 'NEWRELIC:NEW_RELIC_API_KEY';
+
+
+    /**
+     * Specification:
+     * - NewRelic Application ID array
```
Comment (stereomon) [inline src/SprykerEco/Shared/NewRelic/NewRelicEnv.php]: Add a full stop (.) please.

File: src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php (hunk 5 above)
Comment (stereomon) [reply src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php:56]: It makes it a major IMO.

File: composer.json [hunk 9]
```diff
@@ -4,6 +4,7 @@
   "license": "proprietary",
   "require": {
     "php": ">=7.1",
+    "spryker/monitoring": "^2.0.0",
```
Comment (stereomon) [inline composer.json:7]: Why is this now required?

**Review by stereomon — CHANGES_REQUESTED** [review]

File: composer.json (hunk 9 above)
Comment (alex-galych) [reply composer.json:7]: I think this is a mistake.

File: src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php (hunk 5 above)
Comment (fciacchi) [reply src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php:56]: Hi @stereomon - since I'm not in Spryker, I won't be able to be the one releasing or fixing 😛 
However, even if the current stable version is listing this and next parameter optional, requires the next one (that is the third one), otherwise NewRelic API call will fail. 
So technically could be a major, but in reality if someone has the command in the cronjob, it has to have up to the third parameter to work properly, therefore changing to REQUIRED makes it in code what already happens in reality (I think really few customers will call parameters directly, therefore having the command failing).

File: composer.json (hunk 9 above)
Comment (fciacchi) [reply composer.json:7]: It's not a mistake. `spryker/monitoring` contains fixes that are strongly recommended to make this package work properly. Also technically this package can't work without `spryker/monitoring`, therefore listing as dependency will force composer to install it alongside this one.

File: src/SprykerEco/Shared/NewRelic/NewRelicEnv.php (hunk 8 above)
Comment (alex-galych) [reply src/SprykerEco/Shared/NewRelic/NewRelicEnv.php]: Fixed.

File: src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php (hunk 6 above)
Comment (stereomon) [reply src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php:62]: ok

File: src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php (hunk 5 above)
Comment (stereomon) [reply src/SprykerEco/Zed/NewRelic/Communication/Console/RecordDeploymentConsole.php:56]: ok

**Review by stereomon — APPROVED** [review]
